src/main/java/org/ambraproject/wombat/config/theme/FilesystemThemeSource.java

```diff
@@ -0,0 +1,160 @@
+import org.yaml.snakeyaml.Yaml;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * A theme source consisting of a directory on the server's local file system.
+ * <p>
+ * A file named {@code sites.yaml}, at the directory's root level, defines the sites. Any file named {@code theme.yaml}
+ * that appears in a subdirectory defines that subdirectory as a theme; the content of the YAML file indicates the
+ * theme's key and its parents.
+ */
+public final class FilesystemThemeSource implements ThemeSource<FileTheme> {
+
+  private final File root;
+
+  public FilesystemThemeSource(File root) {
+    Preconditions.checkArgument(root.isDirectory());
+    this.root = root;
+  }
+
+  private static Map<?, ?> readConfigYamlMap(File file) {
+    try (Reader reader = new BufferedReader(new InputStreamReader(new FileInputStream(file), Charsets.UTF_8))) {
+      return new Yaml().loadAs(reader, Map.class);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public List<Map<String, ?>> readSites() {
+    File sitesFile = new File(root, "sites.yaml");
+    if (!sitesFile.exists()) {
+      // To support multiple theme sources, of which only one has a sites.yaml file, return an empty list.
+      // To check whether all sources are missing sites.yaml, we need to do it higher up.
+      return ImmutableList.of();
+    }
+
+    Map<?, ?> map = readConfigYamlMap(sitesFile);
+    return (List<Map<String, ?>>) map.get("sites");
+  }
+
+  /**
+   * @return a collection of all {@code theme.yaml} files found in this object's directory
+   */
+  private Collection<File> findAllThemeFiles() throws IOException {
+    Collection<File> hits = Collections.synchronizedCollection(new ArrayList<>());
+    Files.walkFileTree(root.toPath(), new SimpleFileVisitor<Path>() {
+      @Override
+      public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+        if (file.getFileName().toString().equalsIgnoreCase("theme.yaml")) {
+          hits.add(file.toFile());
+        }
+        return super.visitFile(file, attrs);
+      }
+    });
+    return hits;
+  }
+
+  private static ThemeBuilder.ConstructorFunction<FileTheme> createConstructorFunction(File themeFile) {
+    File themeDirectory = Objects.requireNonNull(themeFile.getParentFile());
+    return (key, parentObjects) -> new FileTheme(key, parentObjects, themeDirectory);
+  }
```

Comment:
Regarding https://github.com/PLOS/wombat/pull/1254#discussion_r101414286, this is a practical example of how a `ConstructorFunction` plugs into a `ThemeBuilder`. Here, it seems pretty straightforward IMHO.

The one awkward part is how the key is passed to the `ThemeBuilder` constructor, which then passes it back to this `ConstructorFunction`. Tucking the lambda away in this `createConstructorFunction` method actually sweeps this blemish under the rug pretty nicely -- if you were to inline `createConstructorFunction` into `parseThemeDescription`, you can see how `key` would become overloaded. I experimented with a few different ways to move the key value around and this one seemed the least brittle. I like doing it this way because we create the `ConstructorFunction` seeing only the `File` -- i.e., the thing that this `FileTheme` uniquely depends on that wouldn't apply to other `Theme` subclasses.